compiler/quilt/tools/compat.py

```diff
@@ -0,0 +1,63 @@
+"""
+Python 2 / 3 compatibility
+
+In cases where the `six` lib covers an item, use `six`.
+
+For others, import the module here under the python3 name (if there are naming differences).
+
+Unfortunately, this *doesn't* allow for `from .tools.compat.foo import bar`, so if needed,
+import frequently-used objects here for convenience.
+"""
+# unused imports -- pylint: disable=W0611
+
+import sys as _sys
+
+
+# Reflecting requirements in setup.py
+# Python < 3.4
+if _sys.version_info < (3, 4):
+    from funcsigs import signature  # inspect.argspec is deprecated
+else:
+    from inspect import signature   # inspect.argspec is deprecated
+
+# Python < 3.5
+if _sys.version_info < (3, 5):
+    import pathlib2 as pathlib
+else:
+    import pathlib
+
+
+# Example convenience references to allow `from .tools.compat import some_obj`
+
+# patch = mock.patch
+# Path = pathlib.Path
+# TemporaryDirectory = tempfile.TemporaryDirectory
+
+
+# Manual backport of a stdlib tool -- avoid extra backport dependency.
+try:
+    from tempfile import TemporaryDirectory
+except ImportError:
+    import tempfile as _tempfile
+    import contextlib as _contextlib
+    from shutil import rmtree as _rmtree
+
+    @_contextlib.contextmanager
+    def TemporaryDirectory(suffix='', prefix='tmp', dir=None):
```

Comment:
This seems like overkill; same for chmod_context. They're only used in unit tests, right? We already create a temporary directory for each test - so you can create whatever subdirectories you want, chmod them, etc., and not worry about cleanup.

Reply:
Ah, ok.  I was confused because `__file__` is still the specific file in the working tree, so if you do:
```
mydir = os.path.abspath(__file__)
<do stuff with mydir>
```
..it affects the working tree.

Re: chmod_context:  Sounds good.